go/pulldown/pulldown.go

```diff
@@ -0,0 +1,176 @@
+/*
+	Copyright 2022 Loophole Labs
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		   http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package pulldown
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/loopholelabs/scalefile"
+)
+
+const POLICY_PULL_ALWAYS = "always"
+const POLICY_PULL_IF_NOT_PRESENT = "ifNotPresent"
+const POLICY_PULL_NEVER = "never"
+
+var ERROR_HASH_MISMATCH = errors.New("Hash mismatch")
+var ERROR_NO_FUNCTION = errors.New("Function does not exist and pull policy was never")
+
+// Specifies a pulldown request
+type PulldownConfig struct {
+	APIKey       string
+	Organization string
+	Function     string
+	Tag          string
+}
+
+// StoreConfig
+type StoreConfig struct {
+	CacheDirectory string
+	PullPolicy     string
+}
+
+// Default store config
+var DefaultStoreConfig = StoreConfig{
+	CacheDirectory: "~/.cache/scale/functions",
+	PullPolicy:     "allways",
+}
+
+// What we get back from the API call
+type PulldownResponse struct {
+	URL  string
+	Hash string
+}
+
+// Create a new runtime for a specific scalefile
+func New(pc PulldownConfig, sc StoreConfig) (*scalefile.ScaleFile, error) {
```

Comment:
We should use variadic function options here. 

So you'd have a `Config` type with unexported values for the Cache directory and the pull policy, etc etc. 

Then you'd have a type called `Option` that would be `type Option func(config *Config)`

Then you'd create modifier options like:
```go
func WithPullPolicy(pullPolicy PullPolicy) Option {
  return func(config *Config) {
    config.pullPolicy = pullPolicy
  }
}
```

Then your New function would be `func New(function string, opts ...Options) (*scalefile.ScaleFile, error)`

We know the name of the function is required, but API Key isn't so it'll be a modifier and they'll be able to call it like `New("myorg/myfunc:mytag", WithAPIKey("apikey"), WithPullPolicy(AlwaysPullPolicy))`

Reply:
Removed org and tag, they are now in function as "myorg/myfunc:mytag"

67643bdd6fa13177dc9477d629e185c5d82d29fa